vllm/worker/multi_step_model_runner.py

```diff
@@ -0,0 +1,413 @@
+
+        # Update GPU tensors
+        ops.advance_step(num_seqs=num_seqs,
+                         num_queries=num_queries,
+                         block_size=self.block_size,
+                         input_tokens=model_input.input_tokens,
+                         sampled_token_ids=out.sampled_token_ids,
+                         input_positions=model_input.input_positions,
+                         seq_lens=attn_metadata.seq_lens_tensor,
+                         slot_mapping=attn_metadata.slot_mapping,
+                         block_tables=attn_metadata.block_tables)
+
+        # Update sampling_metadata
+        # model_input.seq_lens = attn_metadata.seq_lens
+        # sampling_metadata = model_input.sampling_metadata
+        # self._update_sampling_metadata(sampling_metadata, num_seqs,
+        #                                num_queries)
+        for i in range(num_queries):
+            model_input.seq_lens[i] = attn_metadata.seq_lens[i]
+
+        return model_input
+
+
+def _pythonize_sampler_output(
+        model_input: ModelInputForGPUWithSamplingMetadata,
+        output: SamplerOutput) -> SamplerOutput:
+    # TODO(will): fix logprobs
+
+    # samples generation should have been skipped
+    assert not output.outputs
+
+    # print shapes
+    # print('output.sampled_token_ids', output.sampled_token_ids.shape)
+    # print('output.logprobs', output.logprobs.shape)
+    samples = torch.empty(*output.sampled_token_ids.shape,
+                          dtype=output.sampled_token_ids.dtype,
+                          device="cpu",
+                          pin_memory=True)
+    # prompt_logprobs = torch.empty(
+    #     *output.prompt_logprobs.shape,
+    #     dtype=output.prompt_logprobs.dtype,
+    #     device="cpu",
+    #     pin_memory=True)
+
+    # CPU GPU sync
+    # logprobs = logprobs.copy_(output.logprobs, non_blocking=False)
+    samples = samples.copy_(output.sampled_token_ids, non_blocking=False)
+
+    samples_list = samples.tolist()
+    del samples
+    # logprobs = logprobs.tolist()
+    # print('samples', samples)
+    # print('logprobs', logprobs)
+
+    # from vllm.model_executor.layers.sampler import _get_logprobs
+
+    # output.sampled_token_ids = output.sampled_token_ids.cpu()
+    # token_ids = output.sampled_token_ids.tolist()
+    sampling_metadata = model_input.sampling_metadata
+
+    for (seq_group, sample_result) in zip(sampling_metadata.seq_groups,
+                                          samples_list):
+        seq_ids = seq_group.seq_ids
+        # next_token_ids, parent_ids = sample_result
+        next_token_ids = sample_result
+        parent_ids = [0]
+        seq_outputs: List[SequenceOutput] = []
+        for parent_id, next_token_id in zip(parent_ids, next_token_ids):
+            # print('SequenceOutput', seq_ids[parent_id], next_token_id)
+            # XXX Hard coded logprob
+            seq_outputs.append(
```

Comment:
This will execute async which will lead to silent data corruption when running `logits_postprocessors` on every model forward.

https://github.com/vllm-project/vllm/blob/997cf78308d292b03c8a1e68d8d1a1f599551937/vllm/model_executor/layers/logits_processor.py#L128C62-L128C78

Reply:
Please add assert checks and forbid config that would allow such possibility

Reply:
This method is async but will block for the corresponding forward to complete using cuda events.

Reply:
I think you misunderstand me. The consumer of the output of this async append (i.e. `logits_processor`) will read an inconsistent unless you explicitly synchronize before it reads.

So this means that prior to a model forward, if there is any `logits_processor`, you need to run `pythonize` rather than `maybe_pythonize`.

Reply:
Ah I understand now. Good catch! added an assert and message. I think for supporting this in the future and support other scenarios (ie guided decoding) we can have a flag that enforce synchronous pythonization with multi-step, at the cost of some tput.